ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/batch/BatchedSubmissionValidator.scala

```diff
@@ -0,0 +1,439 @@
+import com.codahale.metrics.{Counter, Histogram, Timer}
+import com.daml.ledger.participant.state.kvutils.DamlKvutils._
+import com.daml.ledger.participant.state.kvutils.api.LedgerReader
+import com.daml.ledger.participant.state.kvutils.{Envelope, KeyValueCommitting}
+import com.daml.ledger.participant.state.v1.ParticipantId
+import com.daml.ledger.validator
+import com.daml.ledger.validator.SubmissionValidator.LogEntryAndState
+import com.daml.ledger.validator._
+import com.daml.lf.data.Time
+import com.daml.lf.data.Time.Timestamp
+import com.daml.lf.engine.Engine
+import com.daml.logging.LoggingContext.newLoggingContext
+import com.daml.logging.{ContextualizedLogger, LoggingContext}
+import com.daml.metrics.{MetricName, Metrics, Timed}
+import com.google.protobuf.ByteString
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.{Failure, Success}
+
+object BatchedSubmissionValidator {
+  def apply[CommitResult](
+      params: BatchedSubmissionValidatorParameters,
+      committer: KeyValueCommitting,
+      conflictDetection: ConflictDetection,
+      metrics: Metrics,
+      engine: Engine)(
+      implicit executionContext: ExecutionContext): BatchedSubmissionValidator[CommitResult] =
+    new BatchedSubmissionValidator[CommitResult](
+      params,
+      committer,
+      engine,
+      conflictDetection,
+      metrics
+    )
+
+  private[validator] def apply[CommitResult](
+      params: BatchedSubmissionValidatorParameters,
+      engine: Engine,
+      metrics: Metrics)(
+      implicit executionContext: ExecutionContext): BatchedSubmissionValidator[CommitResult] =
+    new BatchedSubmissionValidator[CommitResult](
+      params,
+      new KeyValueCommitting(engine, metrics),
+      engine,
+      new ConflictDetection(metrics),
+      metrics)
+
+  private type CorrelationId = String
+
+  /** A [[DamlSubmission]] with an associated correlation id and a log entry id computed
+    * from the envelope. */
+  private case class CorrelatedSubmission(
+      correlationId: CorrelationId,
+      logEntryId: DamlLogEntryId,
+      submission: DamlSubmission)
+
+  private val LogEntryIdPrefix = "0"
+
+  // While the log entry ID is no longer the basis for deriving absolute contract IDs,
+  // it is used for keying log entries / fragments. We may want to consider content addressing
+  // instead and remove the whole concept of log entry identifiers.
+  // For now this implementation uses a sha256 hash of the submission envelope in order to generate
+  // deterministic log entry IDs.
+  private def bytesToLogEntryId(bytes: ByteString): DamlLogEntryId = {
+    val messageDigest = MessageDigest
+      .getInstance("SHA-256")
+    messageDigest.update(bytes.asReadOnlyByteBuffer())
+    val hash = messageDigest
+      .digest()
+      .map("%02x" format _)
+      .mkString
+    val prefixedHash = ByteString.copyFromUtf8(LogEntryIdPrefix + hash)
```

Comment:
It's possible but didn't seem to be worth doing, i.e., the log entry ID has fixed length (64 bytes) whereas the submission may be significantly larger (and we don't copy that). `ByteString` does not have a public implementation that would allow working off of a mutable byte array.